connector/countconnector/counter.go

```diff
@@ -117,3 +125,24 @@ func (c *counter[K]) appendMetricsTo(metricSlice pmetric.MetricSlice) {
 		}
 	}
 }
+
+func getNestedAttribute(attr string, depth int, attributes pcommon.Map) (pcommon.Value, bool) {
+	logger := zap.NewExample()
+	defer logger.Sync()
+	left, right, _ := strings.Cut(attr, ".")
```

Comment:
I think we're still misaligned here. We shouldn't need to split any strings.

Reply:
Sure. I may definitely be misunderstanding or have lost the thread. Are you saying stop cutting and doing the recursion entirely and do this a different way? Or what would you suggest doing instead?

Appreciate your thoughts and knowledge here!

Reply:
Apologies for delayed response, just getting caught up after some time away.

I think we should _not_ be cutting strings in any case, but we _should_ be recursing through nested maps.

We need a function that looks roughly like this:
```go
func countNestedAttribute(attr string, maxDepth int, attributes pcommon.Map) int {
    if maxDepth == 0 {
        return 0 // done recursing
    }
    var count int
    // Range will iterate over key/value pairs in the map.
    // The bool indicates whether or not to keep iterating.
    attributes.Range(func(k string, v pcommon.Value) bool {
        if k == attr {
            count++
            return true
        }
        if v.Type() != pcommon.ValueTypeMap {
            return true
        }
        count += countNestedAttribute(attr, maxDepth-1, v.Map())
        return true
    })
    return count
}
```

Using an example input:
```yaml
attributes:
  foo: 1
  not.foo: 0 # a different attribute, unrelated to 'foo'
  other: # a nested map that contains additional attributes
    foo: 2
```
If we call `countNestedAttribute("foo", 100, attributes)`, I'd expect the following to happen:
- We find "foo" at the top level and count +1
- We see "not.foo", but this is not the attribute we are looking for so we ignore it
- We see that "other" is a map, so we recurse by calling `countNestedAttribute("foo", 99, attributes["other"])`
  - Within the recursed call, we find "foo", so we count +1
- The end result is 2

Reply:
Ok that makes sense and clarifies perfectly for me!
Appreciate the detailed reply and hope your time away was good!

Reply:
After looking through this I feel like I may actually be more confused. 

The `getNestedAttribute` function in question (where the string cutting is happening) doesn't produce a count/int it produces the string value to fill in for the given attribute key. The counting of that key/value pair happens elsewhere.

Are you saying I should be doing the actual counting here or am I misunderstanding the meaning of count in this case?

Maybe the example attribute values you reference being ints is also confusing here? I'd expect them to be strings (or maybe ints as strings)?

My example data (i'm essentially trying to get out `oof` and if blank i'd give the `default_value`):
```
          - attributes:
              - key: other
                value:
                  kvlistValue: 
                    values:
                      - key: foo
                        value:
                           stringValue: oof
```

Am I totally confused here?

Reply:
I only meant to clarify _how_ attributes should be counted.

Reply:
I guess what I'm not clear on is how that counting replaces the string cutting to get the attribute stringvalue out of the array?

It feels like we may be talking about two different things which is why I'm asking for clarification. I'm sorry if this sounds daft but I'm just really unclear on what is being asked for exactly.

Thanks! Appreciate your feedback!

Reply:
Hi Jeremy,

I don't think we can make progress on this issue until we agree that there is a difference between the following:
1) an attribute whose _value_ is an map that contains other attributes
2) an attribute whose _key_ contains one or more `.`

```
attributes:
  hello:
    foo: oof
  hello.foo: oof
```

---

If we are trying to count attributes called "foo", then only the first case above should be counted.

String splitting is only necessary if you are trying to count "hello.foo" as "foo", which is incorrect.

Reply:
That approach doesn't actually provide an answer though as it isn't returning the value from the `kvlist`. It simply returns a count of how many times the recursion in the function happened which doesn't help with getting the value string out of the nested key.

Or am I totally missing something here?

Reply:
In the mean time I'm doing this in the transform processor to be able to count nested fields
`- set(attributes["nested.attribute.foo"], attributes["nested"]["attribute"]["foo"]) where attributes["nested"]["attribute"]["foo"] != nil`

Which turns the nested key and value into a single named dot delimited key and associated value that isn't buried in a map/`kvlist`

Reply:
@greatestusername, if we agree that `attributes["nested.attribute.foo"]` is different than `attributes["nested"]["attribute"]["foo"]`, then I don't understand why the solution would involve tokenizing `"nested.attribute.foo"`. What am I missing here?

Reply:
I think the primary confusion is around not being able to reference `attributes["nested"]["attribute"]["foo"]` directly in the config and having to use the "made up" string of `nested.attribute.foo`?

If I could reference it as `attributes["nested"]["attribute"]["foo"]` in the config rather than `nested.attribute.foo` there wouldn't be an issue as we'd be able to reference the nested maps explicitly. 

```
  count:
    logs:
      logs.to.metric:
        description: turning logs into metrics
        attributes:
          - key: attributes["nested"]["attribute"]["foo"]
            default_value: None 
```
Would then describe we want the value of `foo` under `attribute` map which is under `nested` map. 

But as it stands the only config I can make stick is:
```        
  count:
    logs:
      logs.to.metric:
        description: turning logs into metrics
        attributes:
          - key: nested.attribute.foo
            default_value: None 
```
Which I think is causing a lot of the confusion in this thread because `nested.attribute.foo` is "made up" and not an actual OTTL description of the data path?

In other words would it make more sense to have the `- key:` config accept standard OTTL syntax?

Reply:
@greatestusername, thanks for detailing this out. I agree with where you're going here. Let's use OTTL syntax for referring to nested attributes.

Reply:
Would this be a larger change to the connector than initially proposed? I.E. would this change the config entirely to always expect OTTL? in the `key` part of the config? Regardless of if it was a nested or non-nested attribute?

Before doing this work I'd want to outline with you your expectations and what you'd like to see as it is a larger set of changes I'd imagine than the issue initially scoped.

What are your thoughts on this?

Apologies for the couple days without a response. Life took over.
Thank you again for your help, review, and patience!

Reply:
@greatestusername, it's not immediately clear to me what the solution should be. Perhaps it would would be helpful to include an example config that shows a possible solution.

Reply:
This config from my example above would theoretically be what I'd expect for the example data I've been discussing if `key` accepted and used `OTTL` [syntax/conventions](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/pkg/ottl/README.md#paths).
```
  count:
    logs:
      logs.to.metric:
        description: turning nested attribute into metric
        attributes:
          - key: attributes["nested"]["attribute"]["foo"]
            default_value: None 
```

Essentially that is just using the OTTL conventions in the `key` section of the config.

That is what brings up my previous question though. Should `key` be changed to always be expecting OTTL syntax/conventions generally?

In that case getting an attribute like `single.dot.named.key` would be configured like 
```
  count:
    logs:
      logs.to.metric:
        description: turning attribute into count
        attributes:
          - key: attributes["single.dot.named.key"]
            default_value: None 
```
Where currently the config for `key` would just be `- key: single.dot.named.key` (not using OTTL syntax).

In other words, should this be a larger change to the connector (using OTTL syntax for all `key` config attributes)?

Appreciate your continued thoughts and review on this!

Reply:
Thanks for expanding this for me.

I see why it would require separate formats, but I think the `attributes["..."]` framing would seem extraneous to most users who only care about non-nested attributes.

My opinion is that it would be reasonable to support both formats for now. Essentially by adding the `attributes["..."]` wrapper automatically to `single.do.named.key`. WDYT?

Reply:
So the "backend" would always be using OTTL syntax ("essentially by adding the `attributes["..."]` wrapper automatically")?

I.E. The connector would do a check on `key` from the config and if it didnt have `attributes[]` syntax it would add that before doing the OTTL evaluation on any `key` provided?

Thus `-key: single.dot.named.key` would be changed to `attributes["single.dot.named.key"]` inside of the connector internal functions. 
But `-key: attributes["single.dot.named.key"]` would not be changed because it is already in the OTTL syntax and get used as written.


Just want to verify that I'm understanding what you're saying the way you're saying it.

Reply:
I think that makes sense. WDYT?

Reply:
That seems reasonable to me as a solution to minimize any breakage to current users.

Reply:
Just wanted to confirm this is the path forward you'd like to see before starting in on it. We good? 😃

Reply:
Yes, I believe we're on the same page.